### ISSUE TITLE
Incorrect solutions given by `solve` 

### ISSUE THREAD

**Issue by mmsbrggr** [sympy/sympy#21882]
The following code tries to find the solutions of a system of (multivariate quadratic) equations. However, the single solution returned by `solve` is not correct.
It seems like `f = 0` should be part of the solution but is not.

```python
from sympy import solve, sympify

equations = [
    "-k*a + b + 5*f/6 + 2*c/9 + 5*d/6 + 4*a/3",
    "-k*f + 4*f/3 + d/2",
    "-k*d + f/6 + d",
    "13*b/18 + 13*c/18 + 13*a/18",
    "-k*c + b/2 + 20*c/9 + a",
    "-k*b + b + c/18 + a/6",
    "5*b/3 + c/3 + a",
    "2*b/3 + 2*c + 4*a/3",
    "-g"
]
equations = [sympify(e) for e in equations]
unknowns = {s for e in equations for s in e.free_symbols}

solutions = solve(equations, unknowns, dict=True)
# Should all be zero but are not:
for solution in solutions:
    for e in equations:
        print(e.subs(solution).simplify())

```

Output:
```
0
f*(-12*k**2 + 28*k - 15)/(12*(k - 1))
0
65*f*(5 - 6*k)/(108*(3*k**2 - 5*k + 2))
5*f*(6*k - 5)*(9*k - 17)/(36*(k - 1)*(3*k - 2))
0
0
35*f*(5 - 6*k)/(18*(3*k**2 - 5*k + 2))
0
```

Expected Output:
```
0
0
0
0
0
0
0
0
0
```

The bug exists in sympy versions `1.8` and `1.7` but NOT in version `1.6`.

In the meantime, is there a way to circumvent this problem until the issue is fixed? We thought about calling `solve` again on the equations of the output but are not sure if that is the best way.

**Comment by oscarbenjamin** [comment]
I've bisected this to 67fbbe670dbf2e77b4fdebf6493ee4f32f9c8a79 from #18844.

A simple diff that fixes this is:
```diff
diff --git a/sympy/solvers/solvers.py b/sympy/solvers/solvers.py
index fd9e6a5fc8..aea9652aa2 100644
--- a/sympy/solvers/solvers.py
+++ b/sympy/solvers/solvers.py
@@ -1737,7 +1737,7 @@ def _solve_system(exprs, symbols, **flags):
     if not exprs:
         return []
 
-    if flags.pop('_split', True):
+    if flags.pop('_split', False):
         # Split the system into connected components
         V = exprs
         symsset = set(symbols)
```
However it should be possible to split the system so there must be a bug somehow that was introduced by the changes in that commit.

**Comment by oscarbenjamin** [comment]
This is a better fix but I'm not quite sure why it works:
```diff
diff --git a/sympy/solvers/solvers.py b/sympy/solvers/solvers.py
index fd9e6a5fc8..dbe97dd12a 100644
--- a/sympy/solvers/solvers.py
+++ b/sympy/solvers/solvers.py
@@ -1757,18 +1757,7 @@ def _solve_system(exprs, symbols, **flags):
                 subsyms = set()
                 for e in subexpr:
                     subsyms |= exprsyms[e]
-                subsyms = list(sorted(subsyms, key = lambda x: sym_indices[x]))
-                # use canonical subset to solve these equations
-                # since there may be redundant equations in the set:
-                # take the first equation of several that may have the
-                # same sub-maximal free symbols of interest; the
-                # other equations that weren't used should be checked
-                # to see that they did not fail -- does the solver
-                # take care of that?
-                choices = sift(subexpr, lambda x: tuple(ordered(exprsyms[x])))
-                subexpr = choices.pop(tuple(ordered(subsyms)), [])
-                for k in choices:
-                    subexpr.append(next(ordered(choices[k])))
+                subsyms = sorted(subsyms, key=default_sort_key)
                 flags['_split'] = False  # skip split step
                 subsol = _solve_system(subexpr, subsyms, **flags)
                 if not isinstance(subsol, list):
```
Apparently that comment was committed by me in 67fbbe670dbf2e77b4fdebf6493ee4f32f9c8a79 but the code in that commit doesn't look like something that I would write and I don't understand what the comment means...

@smichr do you understand what this is doing?

**Comment by oscarbenjamin** [comment]
> This is a better fix but I'm not quite sure why it works:

That diff gives three test failures in the solvers tests:
```
test_issue_5849
test_issue_5849_matrix
test_issue_18208
```
I think that in the the #5849 case the tests might just be wrong. #18208 seems related and made a change to the ordering here. Actually seeing that a better fix here is:
```diff
diff --git a/sympy/solvers/solvers.py b/sympy/solvers/solvers.py
index fd9e6a5fc8..acb12eb421 100644
--- a/sympy/solvers/solvers.py
+++ b/sympy/solvers/solvers.py
@@ -1758,17 +1758,6 @@ def _solve_system(exprs, symbols, **flags):
                 for e in subexpr:
                     subsyms |= exprsyms[e]
                 subsyms = list(sorted(subsyms, key = lambda x: sym_indices[x]))
-                # use canonical subset to solve these equations
-                # since there may be redundant equations in the set:
-                # take the first equation of several that may have the
-                # same sub-maximal free symbols of interest; the
-                # other equations that weren't used should be checked
-                # to see that they did not fail -- does the solver
-                # take care of that?
-                choices = sift(subexpr, lambda x: tuple(ordered(exprsyms[x])))
-                subexpr = choices.pop(tuple(ordered(subsyms)), [])
-                for k in choices:
-                    subexpr.append(next(ordered(choices[k])))
                 flags['_split'] = False  # skip split step
                 subsol = _solve_system(subexpr, subsyms, **flags)
                 if not isinstance(subsol, list):
```
I'm still not sure what the removed code does. With that diff though `test_issue_18208` passes but the other two for #5849 still fail.

**Comment by oscarbenjamin** [comment]
Fix plus update to tests:
```diff
diff --git a/sympy/solvers/solvers.py b/sympy/solvers/solvers.py
index fd9e6a5fc8..acb12eb421 100644
--- a/sympy/solvers/solvers.py
+++ b/sympy/solvers/solvers.py
@@ -1758,17 +1758,6 @@ def _solve_system(exprs, symbols, **flags):
                 for e in subexpr:
                     subsyms |= exprsyms[e]
                 subsyms = list(sorted(subsyms, key = lambda x: sym_indices[x]))
-                # use canonical subset to solve these equations
-                # since there may be redundant equations in the set:
-                # take the first equation of several that may have the
-                # same sub-maximal free symbols of interest; the
-                # other equations that weren't used should be checked
-                # to see that they did not fail -- does the solver
-                # take care of that?
-                choices = sift(subexpr, lambda x: tuple(ordered(exprsyms[x])))
-                subexpr = choices.pop(tuple(ordered(subsyms)), [])
-                for k in choices:
-                    subexpr.append(next(ordered(choices[k])))
                 flags['_split'] = False  # skip split step
                 subsol = _solve_system(subexpr, subsyms, **flags)
                 if not isinstance(subsol, list):
diff --git a/sympy/solvers/tests/test_solvers.py b/sympy/solvers/tests/test_solvers.py
index ef5a4176c7..9c9ea50e81 100644
--- a/sympy/solvers/tests/test_solvers.py
+++ b/sympy/solvers/tests/test_solvers.py
@@ -1316,6 +1316,11 @@ def test_issue_5114_solvers():
 
 
 def test_issue_5849():
+    #
+    # XXX: This system does not have a solution for most values of the
+    # parameters. Generally solve returns the empty set for systems that are
+    # generically inconsistent.
+    #
     I1, I2, I3, I4, I5, I6 = symbols('I1:7')
     dI1, dI4, dQ2, dQ4, Q2, Q4 = symbols('dI1,dI4,dQ2,dQ4,Q2,Q4')
 
@@ -1332,21 +1337,24 @@ def test_issue_5849():
     )
 
     ans = [{
-    I1: I2 + I6,
-    dI1: -4*I2 - 4*I3 - 4*I5 - 10*I6 + 24,
-    I4: -I5 + I6,
-    dQ4: -I5 + I6,
-    Q4: 3*I5/2 - I6/2 - dI4/2,
+    I1: I2 + I3,
+    dI1: -4*I2 - 8*I3 - 4*I5 - 6*I6 + 24,
+    I4: I3 - I5,
+    dQ4: I3 - I5,
+    Q4: -I3/2 + 3*I5/2 - dI4/2,
     dQ2: I2,
     Q2: 2*I3 + 2*I5 + 3*I6}]
 
     v = I1, I4, Q2, Q4, dI1, dI4, dQ2, dQ4
     assert solve(e, *v, manual=True, check=False, dict=True) == ans
-    assert solve(e, *v, manual=True) == ans[0]
+    assert solve(e, *v, manual=True, check=False) == ans[0]
+    assert solve(e, *v, manual=True) == []
+    assert solve(e, *v) == []
+
     # the matrix solver (tested below) doesn't like this because it produces
     # a zero row in the matrix. Is this related to issue 4551?
     assert [ei.subs(
-        ans[0]) for ei in e] == [-I3 + I6, I3 - I6, 0, 0, 0, 0, 0, 0, 0]
+        ans[0]) for ei in e] == [0, 0, I3 - I6, -I3 + I6, 0, 0, 0, 0, 0]
 
 
 def test_issue_5849_matrix():
@@ -1373,14 +1381,7 @@ def test_issue_5849_matrix():
         2*I3 + 2*I5 + 3*I6 - Q2,
         I4 - 2*I5 + 2*Q4 + dI4
     )
-    assert solve(e, I1, I4, Q2, Q4, dI1, dI4, dQ2, dQ4) == {
-    I1: I2 + I6,
-    dI1: -4*I2 - 4*I3 - 4*I5 - 10*I6 + 24,
-    I4: -I5 + I6,
-    dQ4: -I5 + I6,
-    Q4: 3*I5/2 - I6/2 - dI4/2,
-    dQ2: I2,
-    Q2: 2*I3 + 2*I5 + 3*I6}
+    assert solve(e, I1, I4, Q2, Q4, dI1, dI4, dQ2, dQ4) == []
 
 
 def test_issue_5901():
```

**Comment by oscarbenjamin** [comment]
I've opened a PR for this: #21883

**Comment by mmsbrggr** [comment]
@oscarbenjamin We tried out your solution and it seems to resolve the issue mentioned at the beginning but introduces a new problem: Renaming the symbols appearing in the equations changes the solutions (also the number of solutions) we get. For example the following code solves two identical systems then just contains the symbol "_k" instead of "k":

```python
from sympy import solve, sympify, Symbol

equations = [
    "-k*a + b + 5*f/6 + 2*c/9 + 5*d/6 + 4*a/3",
    "-k*f + 4*f/3 + d/2",
    "-k*d + f/6 + d",
    "13*b/18 + 13*c/18 + 13*a/18",
    "-k*c + b/2 + 20*c/9 + a",
    "-k*b + b + c/18 + a/6",
    "5*b/3 + c/3 + a",
    "2*b/3 + 2*c + 4*a/3",
    "-g"
]
equations = [sympify(e) for e in equations]
unknowns = {s for e in equations for s in e.free_symbols}
solutions = solve(equations, unknowns, dict=True)
print(solutions)

equations = [e.subs({Symbol("k"): Symbol("_k")}) for e in equations]
unknowns = {s for e in equations for s in e.free_symbols}
solutions = solve(equations, unknowns, dict=True)
print(solutions)

```

Output:
```
[{a: 0, d: 0, b: 0, f: 0, c: 0, g: 0}, {a: 0, d: -f, b: 0, k: 5/6, c: 0, g: 0}, {a: -2*c, d: 0, b: c, f: 0, k: 13/18, g: 0}, {a: 0, d: 0, b: 0, k: 3/2, c: 0, f: 0, g: 0}]
[{a: 0, _k: 5/6, b: 0, d: -f, c: 0, g: 0}, {a: 0, _k: 3/2, b: 0, d: 0, c: 0, f: 0, g: 0}]
```

**Comment by oscarbenjamin** [comment]
> We tried out your solution and it seems to resolve the issue mentioned at the beginning but introduces a new problem:

When you say that this introduces a new problem what do you mean?

I get the same output you showed when testing with master so I don't think it is caused by the change I suggested but is rather a different issue with `solve`. With master:
```
[{a: 0, d: 0, b: 0, f: 0, c: 0, g: 0}, {a: 0, d: -f, b: 0, k: 5/6, c: 0, g: 0}, {a: -2*c, d: 0, b: c, f: 0, k: 13/18, g: 0}, {a: 0, d: 0, b: 0, k: 3/2, c: 0, f: 0, g: 0}]
[{a: 0, _k: 5/6, b: 0, d: -f, c: 0, g: 0}, {a: 0, _k: 3/2, b: 0, d: 0, c: 0, f: 0, g: 0}]
```

**Comment by mmsbrggr** [comment]
With master I get the following output:
```
[{a: 5*f*(6*k - 5)/(12*(k - 1)*(3*k - 2)), b: 0, c: -5*f*(6*k - 5)/(4*(k - 1)*(3*k - 2)), d: f/(6*(k - 1)), g: 0}]
[{_k: (6*d + f)/(6*d), a: 5*d*(d + f)/(2*d + f), b: 0, c: -15*d*(d + f)/(2*d + f), g: 0}]
```

**Comment by oscarbenjamin** [comment]
> With master I get the following output:

Oh, yes I get that now. I must have gotten confused and thought I was testing master when I wasn't.

There's clearly something fishy going on somewhere but it shouldn't be to do with the diff I showed. Those outputs on master are both incorrect. Let's try to work out what the correct answer should be from the Groebner basis:
```python
In [7]: for eq in equations: print(eq)
-a*k + 4*a/3 + b + 2*c/9 + 5*d/6 + 5*f/6
d/2 - f*k + 4*f/3
-d*k + d + f/6
13*a/18 + 13*b/18 + 13*c/18
a + b/2 - c*k + 20*c/9
a/6 - b*k + b + c/18
a + 5*b/3 + c/3
4*a/3 + 2*b/3 + 2*c
-g

In [8]: for eq in groebner(equations).exprs: print(eq)
a + 2*c
b - c
c*f
c*k - 13*c/18
d + f
f*k - 5*f/6
g
```
So we can see that `g=0`, `d=-f`, `b=c` and `a=-2*c`. The remaining unknowns are `c`, `f` and `k` and the remaining equations are
```python
c*f
c*(k - 13/18)
f*(k - 5/6)
```
Now either `c=0` or `f=0`:

Case 1: `c=0`. The final equation is
```python
f*(k - 5/6)
```
and either `f=0` and `k` is anything (case 1a) or `k=5/6` and `f` is anything (case 1b)

Case 2: `f=0`. The final equation is
```python
c*(k - 13/18)
```
and either `c=0` and `k` is anything (case 2a) or `k=13/18` and `c` is anything (case 2b).

The 4 cases given by solve with the diff above are:
```python
{a: 0, d: 0, b: 0, f: 0, c: 0, g: 0}    # case 1a and 2a (same)
{a: 0, d: -f, b: 0, k: 5/6, c: 0, g: 0}    # case 1b
{a: -2*c, d: 0, b: c, f: 0, k: 13/18, g: 0}  # case 2b
{a: 0, d: 0, b: 0, k: 3/2, c: 0, f: 0, g: 0}    # other
```
Now the "other" case is a valid solution but it is in fact a special case of the first solution which has the same zeros but allows `k` to be arbitrary - that case could be eliminated without losing anything from the full solution set since it is implied by the first case.

So the correct result to return should be:
```python
{a: 0, d: 0, b: 0, f: 0, c: 0, g: 0}    # case 1a and 2a (same)
{a: 0, d: -f, b: 0, k: 5/6, c: 0, g: 0}    # case 1b
{a: -2*c, d: 0, b: c, f: 0, k: 13/18, g: 0}  # case 2b
```
Each of these is a valid solution. None are redundant since each solution allows a different symbol ('k', 'f', 'c') to be anything while the other solutions require a particular value for the same symbol. Also I think that the argument above shows that this is a comprehensive set of solutions.

So the diff I showed means that we get correct (although redundant) output for the equations with `k` but not for `k_` so there must be a bug somewhere to do with ordering the symbols. I'm not sure why the output is incorrect on master or why the diff above changes this but there must be a bug.

**Comment by oscarbenjamin** [comment]
This seems to make the results for `k` and `k_` consistent:
```diff
diff --git a/sympy/solvers/solvers.py b/sympy/solvers/solvers.py
index 17540efcc3..a7a6f5beba 100644
--- a/sympy/solvers/solvers.py
+++ b/sympy/solvers/solvers.py
@@ -1886,7 +1886,7 @@ def _ok_syms(e, sort=False):
             rv = (e.free_symbols - solved_syms) & legal
             if sort:
                 rv = list(rv)
-                rv.sort(key=default_sort_key)
+                rv.sort(key=lambda s: s.name.replace('_', ''))
             return rv
 
         solved_syms = set(solved_syms)  # set of symbols we have solved for
```
There must be a bug somewhere in this:
https://github.com/sympy/sympy/blob/801b1045d714686fe1baec6ece94815fe43c003a/sympy/solvers/solvers.py#L1890-L1961

That code is called when `solve_poly_system` fails. I'm not sure why `solve_poly_system` would fail here but there's a clearly a bug in the code to handle the failure.

**Comment by oscarbenjamin** [comment]
So what that code does is it tries to solve the system of equations by solving for one variable at a time. It loops through all equations and then for each equation it checks what unknown symbols are present in the equation. Then the equation is solved as an univariate equation for each symbol in turn. That result is then substituted into the remaining equations and the already found solutions.

The symbols in an equation are tried in the order given by `default_sort_key` so if the equation is `-6*d*k + 6*d + f` then the symbols are tried in the order `d, f, k`. However if `k` is instead `_k` then the symbols are tried in the order `_k, d, f`. That is why using a different symbol name affects the result.

If the equation is e.g. `d + f` then solving for `d` gives `-f` and solving for `f` gives `-d`. Those would give redundant solutions so if any solution to the univariate equation involves symbols that are already solved for then it is discarded. In a case like `d * f` solving for `d` or `f` gives `0` - both solutions will be kept as they do not involve symbols that are already solved for.

The `k = 3/2` solution is generated erroneously because:

1. `-6*d*k + 6*d + f`  is solved for `d` giving  `d = f/(6*(k - 1)`
2. Substituting for `d` in `3*d - 6*f*k + 8*f` leads to `-6*f*k + 8*f + f/(2*(k - 1))`.
3. That is solved for `f` giving `f = 0`.
4. It is also solved for `k` giving `k = 3/2, 5/6`
5. In the case that `k = 3/2` it is subsequently found that `f` must be zero.

The problem is that when `f` is found to be zero the equation at step 2 is no longer valid for determining `k`: substituting `f=0` there gives an equation that does not have `k = 3/2` as a solution. This is not picked up later because the solution with `k=3/2` is a valid solution of the original equation: it is not the most general solution given the values for the other unknowns though. Since the general solution is picked up this one ends up being redundant.

The other problems arise from something similar. For example the solution with `k = 13/18` has `d = f = 0` but solving `-6*_k*d + 6*d + f` for `_k` gives `(d + f/6)/d` which is then `0/0`. A subsequent substitution by `d = f` cancels out before both `d` and `f` are later established to be zero:
```python
In [10]: e.subs({d:0, f:0})
Out[10]: nan

In [11]: e.subs(d, f)
Out[11]: 7/6

In [12]: e.subs(d, f).subs(f, 0)
Out[12]: 7/6
```

Basically the order in which the symbols are solved is suboptimal at best and invalid at worst. In an equation `x*y + z` this code will try to solve for `x` first but it is clearly better to solve for `z` since that can be done without division and therefore without needing to worry about divide by zero.

This code needs to have more careful handling of the domains of the different functions and consider divide by zero more delicately. A minimal improvement would be to solve for symbols in an order determined by how solvable they are e.g. if a symbol appears linearly with a concrete numeric coefficients then that should be preferred over any symbol appearing in a nonlinear term.

**Comment by mmsbrggr** [comment]
Thanks for your effort. I really appreciate it. I would try to solve the issue, but I have to say that I am not that familiar with the internals of sympy and all the mathematical delicacies one needs to include.
Is there, in the meantime, any way to circumvent these problems? The code at the beginning of this issue is just a minimal example I constructed. My application needs to solve systems of polynomial equations (highest degree 2). It's not that I "only" need to solve the system from the example but systems like it (depending on the user input).

**Comment by oscarbenjamin** [comment]
There are lots of ways to make a simple solver that can do this using resultants or Groebner bases. The complicated part is trying to find all cases in an underdetermined system or handling roots of high-degree polynomials: a system of degree 2 polynomials can lead to higher degree univariate polynomials which may not be solvable in radicals.

Do you need all cases (your "expected solution" above suggests not)? Are the solutions always expected to have rational coefficients (you may find this quickly becomes impossible for anything more complicated than the examples shown)?

**Comment by oscarbenjamin** [comment]
This diff seems like a simple improvement for the second problem:
```diff
diff --git a/sympy/solvers/solvers.py b/sympy/solvers/solvers.py
index 17540efcc3..f5c109c0a0 100644
--- a/sympy/solvers/solvers.py
+++ b/sympy/solvers/solvers.py
@@ -1884,9 +1884,22 @@ def _solve_system(exprs, symbols, **flags):
         # be solved.
         def _ok_syms(e, sort=False):
             rv = (e.free_symbols - solved_syms) & legal
+
+            # Solve first for symbols that have lower degree in the equation.
+            # Ideally we want to solve firstly for symbols that appear linearly
+            # with rational coefficients e.g. if e = x*y + z then we should
+            # solve for z first.
+            def key(sym):
+                ep = e.as_poly(sym)
+                if ep is None:
+                    complexity = (S.Infinity, S.Infinity, S.Infinity)
+                else:
+                    coeff_syms = ep.LC().free_symbols
+                    complexity = (ep.degree(), len(coeff_syms & rv), len(coeff_syms))
+                return complexity + (default_sort_key(sym),)
+
             if sort:
-                rv = list(rv)
-                rv.sort(key=default_sort_key)
+                rv = sorted(rv, key=key)
             return rv
 
         solved_syms = set(solved_syms)  # set of symbols we have solved for
```
Using that in combination with the diff above (https://github.com/sympy/sympy/issues/21882#issuecomment-899805657) I get:
```python
In [1]: from sympy import solve, sympify, Symbol
   ...: 
   ...: equations = [
   ...:     "-k*a + b + 5*f/6 + 2*c/9 + 5*d/6 + 4*a/3",
   ...:     "-k*f + 4*f/3 + d/2",
   ...:     "-k*d + f/6 + d",
   ...:     "13*b/18 + 13*c/18 + 13*a/18",
   ...:     "-k*c + b/2 + 20*c/9 + a",
   ...:     "-k*b + b + c/18 + a/6",
   ...:     "5*b/3 + c/3 + a",
   ...:     "2*b/3 + 2*c + 4*a/3",
   ...:     "-g"
   ...: ]
   ...: equations = [sympify(e) for e in equations]
   ...: unknowns = {s for e in equations for s in e.free_symbols}
   ...: solutions = solve(equations, unknowns, dict=True)
   ...: print(solutions)
   ...: 
   ...: equations = [e.subs({Symbol("k"): Symbol("_k")}) for e in equations]
   ...: unknowns = {s for e in equations for s in e.free_symbols}
   ...: solutions = solve(equations, unknowns, dict=True)
   ...: print(solutions)
[{a: 0, f: 0, b: 0, d: 0, c: 0, g: 0}, {a: 0, f: -d, b: 0, k: 5/6, c: 0, g: 0}, {a: -2*c, f: 0, b: c, d: 0, k: 13/18, g: 0}, {a: 0, f: 0, b: 0, k: 3/2, c: 0, d: 0, g: 0}]
[{a: -2*c, f: 0, b: c, d: 0, _k: 13/18, g: 0}, {a: 0, f: -d, b: 0, _k: 5/6, c: 0, g: 0}, {a: 0, f: 0, b: 0, d: 0, c: 0, g: 0}, {a: 0, f: 0, b: 0, _k: 3/2, c: 0, d: 0, g: 0}]
```
So both the `k` and `_k` systems give the same result although the cases in the solution are listed in a different order. The redundant `k=3/2` case is listed for both systems - a deeper fix is needed for that.

**Comment by smichr** [comment]
> do you understand what this is doing?

based on the comment, it is solving one of potentially multiple equations that share the same set of free symbols. I will look more closely later today.

**Comment by mmsbrggr** [comment]
@oscarbenjamin Thank you. I'll try that quick fix for now. Redundant solutions are not an issue for our purposes. It's just important that we get all solutions and no incorrect ones.

**Comment by smichr** [comment]
> seems to make the results for `k` and `k_` consistent:

I'm not sure this is a good idea since then order of processing would be indeterminant.

**Comment by mmsbrggr** [comment]
I just want to note that the bug does not exist in sympy's fork https://github.com/diofant/diofant

**Comment by oscarbenjamin** [comment]
> I just want to note that the bug does not exist in sympy's fork https://github.com/diofant/diofant

Yes, so the code that I'm suggesting to fix is actually the same in both SymPy and Diofant but this code runs as the final fallback for `solve_system` after `linsolve` and `solve_poly_system` have failed. SymPy's `solve_poly_system` only handles the zero-dimensional case but Diofant's also handles the positive dimensional case. It would definitely be good to improve SymPy's `solve_poly_system` as well but this fallback code would still be needed in cases where there are some non-polynomial terms. I expect that you would see the same bug in Diofant if e.g. one of the symbols was replaced by a non-polynomial expression like `log(x)`.

I'm describing this code as a "fallback" here because these particular cases should ideally be handled by `solve_poly_system` but perhaps rather than viewing this is a fallback we should remember that it is the main handler for arbitrary systems of equations. It should be improved anyway regardless of improvements in `solve_poly_system`.

**Comment by oscarbenjamin** [comment]
> > seems to make the results for `k` and `k_` consistent:
> 
> I'm not sure this is a good idea since then order of processing would be indeterminant.

Why is it indeterminant? Do you mean that it would be non-deterministic?

**Comment by smichr** [comment]
> Do you mean that it would be non-deterministic?

yes

**Comment by oscarbenjamin** [comment]
> > Do you mean that it would be non-deterministic?
> 
> yes

How would it be nondeterministic? We should always get the same output for the same input.

These changes are all implemented in #21883 if anyone wants to try it out.

Basically on master it works like this:
```python
# Get a set of unknown symbols present in eq
def get_unknowns(eq):
    return eq.free_symbols & unknown_symbols

# Sort equations into a uniquely defined order with number of unknown
# symbols as the primary key but otherwise an arbitrary tie-breaker rule
for eq in ordered(equations, lambda eq: len(get_unknowns(eq))): 
    # Loop over the symbols using a deterministic but completely
    # arbitrary order (basically sort by symbol names)
    for sym in sorted(get_unknowns(eq), key=default_sort_key):
        # solve for sym and do substitutions and checks
```
The change I suggested changes the inner sort order so that symbols that are easier to sort for are solved first. Instead of `default_sort_key` we use an approximate measure of how easily the symbol can be solved for and then have `default_sort_key` as the tie breaker:
```python
# Solve first for symbols that have lower degree in the equation.
# Ideally we want to solve firstly for symbols that appear linearly
# with rational coefficients e.g. if e = x*y + z then we should
# solve for z first.
def key(sym):
    ep = e.as_poly(sym)
    if ep is None:
        complexity = (S.Infinity, S.Infinity, S.Infinity)
    else:
        coeff_syms = ep.LC().free_symbols
        complexity = (ep.degree(), len(coeff_syms & rv), len(coeff_syms))
    return complexity + (default_sort_key(sym),)

# Same as master although actually this could be improved:
# The equations could be sorted by how easy they are to solve for a symbol
for eq in ordered(equations, lambda eq: len(get_unknowns(eq))): 
    # Loop over the symbols starting with the trivially solvable ones first
    # but using default_sort_key as the ultimate tie-breaker so that the
    # order is always deterministic.
    for sym in sorted(get_unknowns(eq), key=key):
        # solve for sym and do substitutions and checks
```
The sort key for sorting symbols uses these keys in descending order:
```
1. Degree of the equation as a polynomial in the given symbol (infinite if equation is not polynomial)
2. Number of unknown symbols appearing in the leading coefficient
3. Total number of symbols (including parameters as well as unknowns) in leading coefficient
4. Default sort key as a tie-breaker
```
The first key means that in e.g. `x + y**2 + exp(z)` we prefer to try the symbols in the order `x, y, z` because `x` has lower degree than `y` and the equation is not even polynomial in `z`.

The second key means that in `x + y*z` we prefer to solve for `x`. Strictly this polynomial is linear in any one of the symbols if the others are viewed as parameters but the coefficient of `y` is `z` which is also an unknown so solving for `y` gives `-x/z` which has the unknown `z` in the denominator (possible divide by zero). Much better to solve for `x` first here. The second key rates `y` and `z` equally but these will be disambiguated by `default_sort_key` in the end to give a deterministic order.

The third key means that in `a*x + 2*y` where `x` and `y` are unknowns and `a` is a parameter then we prefer to solve for `y` giving `y = a*x/2` rather than for `x` giving `x = 2*y/a`. Again dividing by a symbolic parameter `a` is a potential divide by zero. It's not as bad as dividing by an unknown that might potentially be solved and found to be zero but it's still not as good as dividing by `2` which is never zero.

If the equation is not even polynomial in a symbol then it is given `(oo, oo, oo)` for the first three keys meaning it will always be sorted after any symbol that the equation is polynomial in.

The fourth key kicks in when two symbols have the same value for all of the previous three keys and simply ensures that the ordering is always well defined regardless of the order of iteration over the set returned by `get_unknowns(eq)`. So the result should still always be deterministic although it does still depend on the names of the symbols (e.g. should the result have `d: f` or `f: d`?).

This is still not optimal though. Ideally we would order the *equations* by how easy it is to extract a fully solved symbol from them and then we should be able to efficiently handle something like #17186.

Also ideally where we do need to divide by some function of the unknowns we should remember that denominator so that we can detect if the particular case is invalid or not (I think that's how we can eliminate the redundant `k=3/2` case).

**Comment by oscarbenjamin** [comment]
@mmsbrggr Please try out the current SymPy master branch after #21883 was merged.
